instrumentation-api/src/test/java/io/opentelemetry/instrumentation/api/caching/CacheTest.java

```diff
@@ -0,0 +1,121 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.caching;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class CacheTest {
+
+  @Nested
+  class StrongKeys {
+    @Test
+    void unbounded() {
+      Cache<String, String> cache = Cache.newBuilder().build();
+
+      CaffeineCache<?, ?> caffeineCache = ((CaffeineCache<?, ?>) cache);
+      assertThat(cache.computeIfAbsent("cat", unused -> "meow")).isEqualTo("meow");
+      assertThat(caffeineCache.keySet()).hasSize(1);
+
+      assertThat(cache.computeIfAbsent("cat", unused -> "bark")).isEqualTo("meow");
+      assertThat(caffeineCache.keySet()).hasSize(1);
+
+      assertThat(cache.computeIfAbsent("dog", unused -> "bark")).isEqualTo("bark");
+      assertThat(caffeineCache.keySet()).hasSize(2);
+      assertThat(cache.computeIfAbsent("cat", unused -> "meow")).isEqualTo("meow");
+    }
+
+    @Test
+    void bounded() {
+      Cache<String, String> cache = Cache.newBuilder().setMaximumSize(1).build();
+
+      CaffeineCache<?, ?> caffeineCache = ((CaffeineCache<?, ?>) cache);
+      assertThat(cache.computeIfAbsent("cat", unused -> "meow")).isEqualTo("meow");
+      assertThat(caffeineCache.keySet()).hasSize(1);
+
+      assertThat(cache.computeIfAbsent("cat", unused -> "bark")).isEqualTo("meow");
+      assertThat(caffeineCache.keySet()).hasSize(1);
+
+      assertThat(cache.computeIfAbsent("dog", unused -> "bark")).isEqualTo("bark");
+      caffeineCache.cleanup();
+      assertThat(caffeineCache.keySet()).hasSize(1);
+      assertThat(cache.computeIfAbsent("cat", unused -> "purr")).isEqualTo("purr");
+    }
+  }
+
+  @Nested
+  class WeakKeys {
+    @Test
+    void unbounded() {
+      Cache<String, String> cache = Cache.newBuilder().setWeakKeys().build();
+
+      CaffeineCache<?, ?> caffeineCache = ((CaffeineCache<?, ?>) cache);
+      String cat = new String("cat");
+      String dog = new String("dog");
+      assertThat(cache.computeIfAbsent(cat, unused -> "meow")).isEqualTo("meow");
+      assertThat(caffeineCache.keySet()).hasSize(1);
+
+      assertThat(cache.computeIfAbsent(cat, unused -> "bark")).isEqualTo("meow");
+      assertThat(caffeineCache.keySet()).hasSize(1);
+
+      assertThat(cache.computeIfAbsent(dog, unused -> "bark")).isEqualTo("bark");
+      assertThat(caffeineCache.keySet()).hasSize(2);
+      assertThat(cache.computeIfAbsent(cat, unused -> "meow")).isEqualTo("meow");
+
+      cat = null;
+      System.gc();
+      // Wait for GC to be reflected.
+      await()
+          .untilAsserted(
+              () -> {
+                caffeineCache.cleanup();
+                assertThat(caffeineCache.keySet()).hasSize(1);
+              });
+      assertThat(cache.computeIfAbsent(dog, unused -> "bark")).isEqualTo("bark");
+      dog = null;
+      System.gc();
```

Comment:
GcFinalizable from guava testlib is great for this as more predictable

Reply:
Thanks for the hint! I've always wondered about this but in practice have yet to see a failure due to only calling `gc` once and continue with this pattern. Looking at it

https://github.com/google/guava/blob/master/guava-testlib/src/com/google/common/testing/GcFinalization.java#L186

I feel as if it would be mostly the same to move `System.gc` into the await() predicate, possibly tweaking the polling interval, which is nice to reduce the number of ways of doing something. Might do one or the other if this flakes in the future.